pkg/provider/provider.go

```diff
@@ -225,34 +229,220 @@ func (p *Provider) GetSecretsStoreObjectContent(ctx context.Context, attrib, sec
+}
+
+/*
+Given a base key vault object and a list of object versions and their created dates, find
+the latest kvObject.ObjectVersionHistory versions and return key vault objects with the
+appropriate alias and version.
+
+The alias is determine by the index of the version starting with 0 at the specified version (or
+latest if no version is specified).
+*/
+func getLatestNKeyVaultObjects(kvObject types.KeyVaultObject, kvObjectVersions types.KeyVaultObjectVersionList) []types.KeyVaultObject {
+	baseFileName := kvObject.GetFileName()
+	objects := []types.KeyVaultObject{}
+
+	sort.Sort(kvObjectVersions)
+
+	// if we're being asked for the latest, then there's no need to skip any versions
+	foundFirst := kvObject.ObjectVersion == "" || kvObject.ObjectVersion == "latest"
+
+	for _, objectVersion := range kvObjectVersions {
+		foundFirst = foundFirst || objectVersion.Version == kvObject.ObjectVersion
+
+		if foundFirst {
+			length := len(objects)
+			newObject := kvObject
+
+			newObject.ObjectAlias = filepath.Join(baseFileName, strconv.Itoa(length))
+			newObject.ObjectVersion = objectVersion.Version
+
+			objects = append(objects, newObject)
+
+			if length+1 > int(kvObject.ObjectVersionHistory) {
+				break
+			}
+		}
+	}
+
+	return objects
+}
+
+func (p *Provider) getKeyVaultObjectVersions(ctx context.Context, kvClient *kv.BaseClient, kvObject types.KeyVaultObject, vaultURL string) (versions types.KeyVaultObjectVersionList, err error) {
+	start := time.Now()
+	defer func() {
+		var errMsg string
 		if err != nil {
-			return nil, err
+			errMsg = err.Error()
+		}
+		p.reporter.ReportKeyvaultRequest(ctx, time.Since(start).Seconds(), kvObject.ObjectType, kvObject.ObjectName, errMsg)
+	}()
+
+	switch kvObject.ObjectType {
+	case types.VaultObjectTypeSecret:
+		return getSecretVersions(ctx, kvClient, vaultURL, kvObject)
+	case types.VaultObjectTypeKey:
+		return getKeyVersions(ctx, kvClient, vaultURL, kvObject)
+	case types.VaultObjectTypeCertificate:
+		return getCertificateVersions(ctx, kvClient, vaultURL, kvObject)
+	default:
+		err := errors.Errorf("Invalid vaultObjectTypes. Should be secret, key, or cert")
+		return nil, wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+	}
+}
+
+func getSecretVersions(ctx context.Context, kvClient *kv.BaseClient, vaultURL string, kvObject types.KeyVaultObject) ([]types.KeyVaultObjectVersion, error) {
+	kvVersionsList, err := kvClient.GetSecretVersions(ctx, vaultURL, kvObject.ObjectName, nil)
+	if err != nil {
+		return nil, wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+	}
+
+	secretVersions := types.KeyVaultObjectVersionList{}
+
+	for notDone := true; notDone; notDone = kvVersionsList.NotDone() {
+		for _, secret := range kvVersionsList.Values() {
+			objectVersion := getObjectVersion(*secret.ID)
+			created := date.UnixEpoch()
+
+			if secret.Attributes != nil {
+				created = time.Time(*secret.Attributes.Created)
```

Comment:
I think we should check `secret.Attributes.Created` is not nil before dereferencing